fsthttp/backend.go

```diff
@@ -0,0 +1,84 @@
+package fsthttp
+
+import (
+	"time"
+
+	"github.com/fastly/compute-sdk-go/internal/abi/fastly"
+)
+
+type TLSVersion uint32
+
+const (
+	TLSVersion1   TLSVersion = 0
+	TLSVersion1_1 TLSVersion = 1
+	TLSVersion1_2 TLSVersion = 2
+	TLSVersion1_3 TLSVersion = 3
+)
+
+type BackendOptions struct {
+	HostOverride        string
+	ConnectTimeout      time.Duration
+	FirstByteTimeout    time.Duration
+	BetweenBytesTimeout time.Duration
+	SSLMinVersion       TLSVersion
+	SSLMaxVersion       TLSVersion
+	CertHostname        string
+	CACert              string
+	Ciphers             string
+	SNIHostname         string
+}
+
+func abiBackendOptions(opts *BackendOptions) *fastly.BackendConfigOptions {
+	var abiOpts fastly.BackendConfigOptions
+
+	if opts == nil {
+		return &abiOpts
+	}
+
+	if opts.HostOverride != "" {
+		abiOpts.HostOverride(opts.HostOverride)
+	}
+
+	if opts.ConnectTimeout != 0 {
+		abiOpts.ConnectTimeout(opts.ConnectTimeout)
+	}
+
+	if opts.FirstByteTimeout != 0 {
+		abiOpts.FirstByteTimeout(opts.FirstByteTimeout)
+	}
+
+	if opts.BetweenBytesTimeout != 0 {
+		abiOpts.BetweenBytesTimeout(opts.BetweenBytesTimeout)
+	}
+
+	if opts.SSLMinVersion != 0 {
+		abiOpts.SSLMinVersion(fastly.TLSVersion(opts.SSLMinVersion))
+	}
+
+	if opts.SSLMaxVersion != 0 {
+		abiOpts.SSLMaxVersion(fastly.TLSVersion(opts.SSLMaxVersion))
+	}
```

Comment:
Because `TLSVersion1` has the zero value, it makes these a little awkward.

What is the default minimum version if not provided?  If it's not TLSv1.0 then it's not possible to set it to that, because it has the zero value.

And not that someone would want to do this, but it's not possible to set TLSv1.0 as the maximum version either.

Reply:
Maybe that's what the `UseSSL` bool is for?  I'll see what the JS and Rust SDKs do.

Reply:
I think the `UseSSL` bool affects whether these apply at all, but I think it's orthogonal to the issue here, which is that the zero value has meaning separate from omission.

Let's suppose the default minimum TLS version is 1.2.  (A reasonably safe choice!)

By way of example, if the code stays as it is,

```
if opts.SSLMinVersion != 0 {
	abiOpts.SSLMinVersion(fastly.TLSVersion(opts.SSLMinVersion))
}
```

then it is not possible to set the minimum version to TLSv1.0.

If it is changed to be,

```
if opts.UseSSL {
	abiOpts.SSLMinVersion(fastly.TLSVersion(opts.SSLMinVersion))
}
```

then by default it overrides the system's default minimum version, lowering it from 1.2 to 1.0 even though that's probably not what the user wants.

I think we need to either:
1. Change the constants so that 0 is "no change" and `TLSVersion_1_0` is 1 (and the others bumped up as well)
2. Switch to using pointers, so `nil` is the "no change" value.  (Ew.)
3. Switch to using a builder pattern for the API

I think option 1 matches the style of the rest of the SDK best.  Since it's a type and constant value defined in the `fsthttp` package, it's not a big deal (although it's a little gross) to decrement the value by 1 before casting it to `fastly.TLSVersion`.  (An alternative would be to `switch` on a conversion between the types, which is more verbose but less gross.)

Reply:
And I'm not sure about the JS SDK but the Rust SDK uses a builder pattern for this (and most other APIs in the SDK)

Reply:
Yeah, I wonder if I should switch this to a builder-type API.  That would solve a bunch of these issues.  Let me think a bit more about this.